R/type-unspecified.R

```diff
@@ -1,20 +1,74 @@
 #' A 1d vector of unspecified type
 #'
-#' This is a [partial type][new_partial] used to represent logical vectors
-#' that only contain `NA`. These require special handling because we want to
-#' allow `NA` to specify missingness without requiring a type.
+#' @description
+#' This is the underlying type used to represent logical vectors that only
+#' contain `NA`. These require special handling because we want to allow logical
+#' `NA` to specify missingness that can be cast to any other type.
+#'
+#' [vec_ptype()] and [vec_ptype2()] convert a logical vector of `NA` into an
+#' empty `<unspecified>` type. This type can combine with any other type.
+#'
+#' [vec_ptype_common()] uses both [vec_ptype()] and [vec_ptype2()] to compute
+#' the common type, but then returns a _finalised_ type using
+#' [vec_ptype_finalise()]. The purpose of `vec_ptype_finalise()` is to turn any
+#' remaining `<unspecified>` types back into `<logical>`, which is the more
+#' useful type for callers of `vec_ptype_common()`.
+#'
+#' `vec_ptype_finalise()` is an S3 generic, but it is extremely rare to need to
+#' write an S3 method for this. Data frames (and data frame subclasses) are
+#' already recursively finalised by the default method. The only time you may
+#' need to write an S3 method for `vec_ptype_finalise()` is if your class
+#' _wraps_ an arbitrary vector that has the potential to be a logical vector
+#' containing only `NA`s. See `ivs::iv()` for an example of this, which wraps
+#' arbitrary `start` and `end` vectors of the same type into a single interval
+#' vector class.
```

Comment:
`unspecified()` and `vec_ptype_finalise()` are now documented together, and I took this opportunity to refresh these docs and write down our thoughts about how these functions interact with `vec_ptype()`, `vec_ptype2()`, and `vec_ptype_common()`.

I like the idea that `vec_ptype_finalise()` is now _only_ about finalization of `unspecified` back to `logical`.